File: medium/2385. Amount of Time for Binary Tree to Be Infected/main.cpp
#include <queue>
#include <unordered_map>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    int amountOfTime(TreeNode* root, int start) {
        std::unordered_map<TreeNode*, TreeNode*> parentMap;
        TreeNode* startNode = findStartNode(root, start, nullptr, parentMap);
        if (startNode == nullptr) return 0;

        std::queue<TreeNode*> queue;
        std::unordered_set<TreeNode*> visited;
        queue.push(startNode);
        visited.insert(startNode);

        int minutes = 0;

        while (!queue.empty()) {
            int levelSize = queue.size();
            for (int i = 0; i < levelSize; ++i) {
                TreeNode* currentNode = queue.front();
                queue.pop();

                // Controlla e aggiungi il nodo genitore
                auto it = parentMap.find(currentNode);
                if (it != parentMap.end() && visited.find(it->second) == visited.end() && it->second != nullptr) {
                    queue.push(it->second);
                    visited.insert(it->second);
                }

                // Aggiungi i figli non infetti
                if (currentNode->left && visited.find(currentNode->left) == visited.end()) {
                    queue.push(currentNode->left);
                    visited.insert(currentNode->left);
                }
                if (currentNode->right && visited.find(currentNode->right) == visited.end()) {
                    queue.push(currentNode->right);
                    visited.insert(currentNode->right);
                }
            }

            if (!queue.empty()) {
                minutes++;
            }
        }

        return minutes;
    }

private:
    TreeNode* findStartNode(TreeNode* node, int start, TreeNode* parent, std::unordered_map<TreeNode*, TreeNode*>& parentMap) {
        if (node == nullptr) return nullptr;
        parentMap[node] = parent;

        if (node->val == start) return node;
        
        TreeNode* leftResult = findStartNode(node->left, start, node, parentMap);
        if (leftResult != nullptr) return leftResult;
        
        return findStartNode(node->right, start, node, parentMap);
    }
};

File: medium/2385. Amount of Time for Binary Tree to Be Infected/main.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int AmountOfTime(TreeNode root, int start) {
        Dictionary<TreeNode, TreeNode> parentMap = new Dictionary<TreeNode, TreeNode>();
        TreeNode startNode = FindStartNode(root, start, null, parentMap);
        if (startNode == null) return 0;

        Queue<TreeNode> queue = new Queue<TreeNode>();
        HashSet<TreeNode> visited = new HashSet<TreeNode>();
        queue.Enqueue(startNode);
        visited.Add(startNode);

        int minutes = 0;

        while (queue.Count > 0) {
            int levelSize = queue.Count; 

            for (int i = 0; i < levelSize; i++) {
                TreeNode currentNode = queue.Dequeue();

                // check the parents in the nodes
                TreeNode parentNode;
                if (parentMap.TryGetValue(currentNode, out parentNode) && parentNode != null && !visited.Contains(parentNode)) {
                    queue.Enqueue(parentNode);
                    visited.Add(parentNode);
                }

                // add infected children
                if (currentNode.left != null && !visited.Contains(currentNode.left)) {
                    queue.Enqueue(currentNode.left);
                    visited.Add(currentNode.left);
                }
                if (currentNode.right != null && !visited.Contains(currentNode.right)) {
                    queue.Enqueue(currentNode.right);
                    visited.Add(currentNode.right);
                }
            }

            if (queue.Count > 0) {
                minutes++;
            }
        }

        return minutes;
    }

    private TreeNode FindStartNode(TreeNode node, int start, TreeNode parent, Dictionary<TreeNode, TreeNode> parentMap) {
        if (node == null) {
            return null;
        }
        parentMap[node] = parent;  // memorize the parent node

        if (node.val == start) {
            return node;
        }

        TreeNode leftResult = FindStartNode(node.left, start, node, parentMap);
        if (leftResult != null) {
            return leftResult;
        }

        return FindStartNode(node.right, start, node, parentMap);
    }
}
